### PR TITLE
Multi core ESLint

### PR DESCRIPTION
**Pull request by potlee**
This splits files and runs eslint on them in separate processes to utilize multiple cores. I am seeing lint times reduced by ~2/3

### CODE REVIEW

**Comment by anthonypt87** [comment]
@Raynos hey Jake. Can you help deep with this? seems like a useful thing to have.

File: index.js [hunk 1]
```diff
@@ -13,6 +13,13 @@ var path = require('path')
 var uniq = require('uniq')
 var cloneDeep = require('clone-deep')
 var deepExtend = require('deep-extend')
+var fs = require("fs")
+var path = require("path")
+var cp = require('child_process')
+var _ = require('lodash')
+
+var NUM_FORKS = 8;
```
Comment (Raynos) [inline index.js]: We can use `os.cpus().length` here instead of hard coding 8.

**Review by Raynos — COMMENTED** [review]
We use an older version of uber-standard, aka v3.7

We CANNOT upgrade to v5.

We need to make a v3.7 branch and make this PR against that branch instead of master.

I am going to verify this branch ( rebased on top of 3.7.1 ) locally.

**Comment by Raynos** [comment]
I rebased it myself for verification purposes ( 3.7-run-eslint-in-parallel ).

**Comment by Raynos** [comment]
```
raynos at raynos-ThinkPad-T440p  
~/uber/realtime-api on master*
$ time npm run lint

> realtime-api@1.0.0 lint /home/raynos/uber/realtime-api
> standard -v --reporter stylish && echo '# linter passed'

# linter passed

real	1m4.789s
user	1m4.504s
sys	0m0.836s
```

```
raynos at raynos-ThinkPad-T440p  
~/uber/realtime-api on master*
$ time npm run lint

> realtime-api@1.0.0 lint /home/raynos/uber/realtime-api
> standard -v --reporter stylish && echo '# linter passed'

# linter passed

real	0m44.694s
user	2m14.900s
sys	0m2.872s
```

Somehow the multi core version takes twice as long but only 45seconds of real time. weird.

**Comment by Raynos** [comment]
```
raynos at raynos-ThinkPad-T440p  
~/uber/realtime-api on master*
$ time npm run lint; time npm run lint; time npm run lint;

> realtime-api@1.0.0 lint /home/raynos/uber/realtime-api
> standard -v --reporter stylish && echo '# linter passed'

# linter passed

real	0m42.665s
user	1m39.196s
sys	0m1.520s

> realtime-api@1.0.0 lint /home/raynos/uber/realtime-api
> standard -v --reporter stylish && echo '# linter passed'

# linter passed

real	0m46.126s
user	1m46.232s
sys	0m1.528s

> realtime-api@1.0.0 lint /home/raynos/uber/realtime-api
> standard -v --reporter stylish && echo '# linter passed'

# linter passed

real	0m46.563s
user	1m45.616s
sys	0m1.684s
```

Tried it again with CPUS - 1 number of forks. it reduces lint time by 25% but also turns my laptop into a helicopter.

I think hogging all the CPUs to run the linter will make a developers laptop ( browser, hipchat, editor, music player ) really slow and unresponsive.

The forking approach might not have enough wins HOWEVER I am only trying a 4 core linux laptop.

**Comment by potlee** [comment]
This is what I am seeing:

```
➜  realtime-api git:(master) ✗ time npm run lint                      

> realtime-api@1.0.0 lint /Users/deeppatel/realtime-api
> standard -v --reporter stylish && echo '# linter passed'

# linter passed
npm run lint  58.24s user 2.78s system 99% cpu 1:01.12 total


➜  realtime-api git:(master) ✗ time npm run lint 

> realtime-api@1.0.0 lint /Users/deeppatel/realtime-api
> standard -v --reporter stylish && echo '# linter passed'

# linter passed
npm run lint  60.61s user 3.39s system 273% cpu 23.403 total

```

I'm on a MBP with an i7 so I used 8 forks (4 cores * 2 threads per core). OSX seems to do a good job at keeping everything else responsive and besides we also run our tests in parallel

**Comment by Raynos** [comment]
@potlee 

How about a compromise, we spawn `os.cpus().length - 1` forks. 

I think reserving one CPU should avoid the laptop helicopter CPU fan issue.

**Comment by Raynos** [comment]
Need to change the base on this branch to 3.7.x ( I created a new branch called 3.7.x ).

**Comment by Raynos** [comment]
Need to rebase on top of 3.7.x ;) to avoid pulling unrelated commits.

File: index.js [hunk 2]
```diff
@@ -11,6 +11,15 @@ var glob = require('glob')
 var parallel = require('run-parallel')
 var path = require('path')
 var uniq = require('uniq')
+var cloneDeep = require('clone-deep')
```
Comment (Raynos) [inline index.js]: cloneDeep & deepExtend are not used, bad merge.

File: index.js [hunk 3]
```diff
@@ -120,19 +129,50 @@ function lintFiles (files, opts, cb) {
     var root = commondir(files)
     editorConfigGetIndent(root, function (err, indent) {
       if (err) return cb(err)
-      var result
-      ESLINT_CONFIG.baseConfig.rules.indent = [2, indent]
```
Comment (Raynos) [inline index.js:124]: ^ this line is important, bad merge.

File: test/api.js [hunk 4]
```diff
@@ -9,3 +14,37 @@ test('api usage', function (t) {
...
+    }, function onFiles(err, result) {
+        assert.error(err, 'no error while linting');
+        assert.equal(typeof result, 'object', 'result is an object');
+        assert.equal(result.errorCount, 181, 'error count 181');
+
+        assert.end();
+    });
+});
+
+test('lintText', function t(assert) {
```
Comment (Raynos) [inline test/api.js]: This new lintText test will fail, bad merge.

File: test/api.js [hunk 5]
```diff
@@ -9,3 +14,37 @@ test('api usage', function (t) {
...
   })
 })
+
+test('api usage', function t(assert) {
+    standard.lintFiles([], {
+        cwd: 'bin'
+    }, function onFiles(err, result) {
+        assert.error(err, 'no error while linting');
+        assert.equal(typeof result, 'object', 'result is an object');
+        assert.equal(result.errorCount, 181, 'error count 181');
```
Comment (Raynos) [inline test/api.js]: This magic number is wrong, bad merge.

**Comment by Raynos** [comment]
Please run `node test/api.js` again.

**Review by Raynos — APPROVED** [review]

**Comment by Raynos** [comment]
lgtm, feel free to land and cut a patch ( `npm version patch; git push origin 3.7.x --tags` )

**Comment by potlee** [comment]
@Raynos I created the tag. Can you publish it to npm?

**Comment by Raynos** [comment]
This PR is not merged ? want to merge it ?

**Comment by Raynos** [comment]
Can we also make sure the v3.7.2 commit is on the 3.7.x branch rather then a floating commit without a branch.

**Comment by Raynos** [comment]
I merged it myself :)

```
raynos at raynos-ThinkPad-T440p  
~/uber/standard on 3.7.x
$ npm info uber-standard versions | grep 3.7.2
  '3.7.2',
```

**Pull request merged**